Ocean3D as a tool script (and OceanSurface3D components for rendering options)

Tattomoosa opened the pull request:
Adding as a draft PR because I had to do a lot more than I expected to get things mostly playing nicely, and I didn't keep my changes very minimal, and there's still a few edge cases to iron out and code to clean up, but I think it's far along enough that it's worth getting your eyes on.

The biggest change is the addition of a component `OceanSurface3D` and subclasses for different rendering methods which I mostly added because I wanted to test while diagnosing issues with QuadTree as a tool script and prepare for optional Terrain3D integration. I think there are also plenty of situations when rendering with the simplest PlaneMeshOceanSurface3D makes the most sense, and simplifying the setup of a surface for an end user is useful.

Another change is that Ocean3D copies the default material instead of using it directly. There were a lot of issues with 2 Ocean3D's affecting the same resource so I'm not sure it should be possible. I have been thinking maybe it should use the resource directly as a separate "initial settings" property but then use a different material to write its changes to, or the various initial settings on the material should be properties of Ocean3D itself that it handles applying to the material properly? Still thinking about this.

I think the last big change is that the QuadTree no longer depends on a scene resource.

Also a callout that I understand now the MultiMesh implementation I made might not actually make sense here - MultiMesh doesn't really have performance benefits until we get to waaaaaaay more meshes, and it doesn't support any mesh resolution differences. I'm keeping it around because I think it may be useful as a piece of a simpler LOD solution to test against the quad tree solution later but performance analysis will need to be done.

There are changes committed to the Example scene but it's not in the best state since I was doing experimentation there. I think it's all working properly, just the settings are not great.

I've added a UvBorder.gdshader for debugging QuadTree behavior (because I made a bit of a mess of it for a minute there, haha...). It can be added as a second pass on the ocean material or applied directly to a quadtree and will render just the edges of the plane to make it really obvious what quadtree is doing.

### TODO before PR
- When Ocean3D editor simulation is on, it should pause when a game instance is running
- QuadTree3D makes all of nodes in-editor but only ever shows the lowest LOD, meaning it can be really slow to open scenes
  - visibility detection doesn't seem to work with the editor camera
- Ocean3D doesn't handle changing the FFT settings correctly while simulating, maybe others, need to test more
- General cleanup, documentation, etc.
  
### Some other thoughts

- QuadTree performance:
  - Working solution in editor that doesn't rely on visibility notification by reading from the camera view directly?
    - Can make this method available during runtime as well and see what's more performant
  - Performance improvements:
    - Instantiate necessary higher detail quads lazily (fairly easy change)
    - See if creating one node then duplicating using instantiation for each other node per LOD level is cheaper. Not finding much documentation on that, it might only be useful for keeping node changes in-sync
    - Re-use high detail quads by moving them around as needed, keep actual mesh count as low as possible (fundamental change)

- The ocean needs to grab a camera for shader calculations anyway. This should be the "source of truth" on what camera needs to be looked at for OceanSurfaces as well as OceanEnvironment (or an OceanEnvironment replacement)

- Infinite ocean surface rendering should be possible by moving the ocean surface rendering mesh(es) with the camera
  - Quick test (not committed) shows that this *kind of* works but is a bit jittery
  - I've done this in a (much worse) ocean shader I built a while back, going to see if I did something there I haven't tried here
  - This was a major consideration behind creating the OceanSurface3D class, as it could handle infinity and maybe other properties that will be common across all surface rendering solutions

Sorry this is all a bit disorganized!

Tattomoosa committed:
OceanEnvironment handles null sky_light, sun_light

Tattomoosa committed:
Made Ocean3D a tool script

Tattomoosa committed:
Bool property to toggle Ocean3D processing in editor

Tattomoosa committed:
Created examples folder, moved existing example scene

Tattomoosa committed:
Add basic setup example

Tattomoosa committed:
Updated BasicSetupExample, removed BasicExample

Tattomoosa committed:
Add OceanSurfaceRenderer3D

Tattomoosa committed:
Clean up debugging artifacts

Tattomoosa committed:
Add SurfaceVisual min/max lod distance defaults

Tattomoosa committed:
Ocean3D duplicates default shader

Tattomoosa committed:
OceanSurfaceRenderer exposes more settings, replaced example scene

Tattomoosa committed:
OceanSurface3D rendering classes

Tattomoosa committed:
QuadTree updates, OceanSurface3D examples, debug utilities

- Fixed issues with QuadTree working properly as a tool script
- Added examples for all OceanSurface3D types for better comparisons
- Added UvBorder debug shader for testing QuadTree3D behavior
- Added example scene to test QuadTree3D behavior

Tattomoosa committed:
Ocean3D uses its camera getter internally

Tattomoosa committed:
Fix QuadTree3D switch camera bug

Tattomoosa commented:
Found a bug with the QuadTree implementation that I've fixed here, but let me know if I should PR it separately real quick since it also happens in `devel`

![Screenshot 2024-02-20 at 9 43 24 PM](https://github.com/tessarakkt/godot4-oceanfft/assets/16693238/34601fec-988f-4b86-a06c-18f2c8fb5134)

You can see the low detail mesh and high detail mesh are rendering on top of each other.

To reproduce:
1. Move the first Example scene camera very far away
2. Switch to camera 2

The fix is to change this bit at line 178
```
    for subquad in _subquads:
        if not subquad.lod_select(cam_pos):
            subquad.mesh_instance.visible = true
```

To

```
    for subquad in _subquads:
        subquad.mesh_instance.visible = false
        if not subquad.lod_select(cam_pos):
            subquad.mesh_instance.visible = true
```

Tattomoosa commented:
There's another issue with changing the camera to the second one when the first is very far away, but I'm not sure it's really possible to fix since I think it has to do with how the simulation simplifies at a distance. Basically all the buoyant objects wind up in the wrong place under or over the surface.

tessarakkt commented:
> The biggest change is the addition of a component OceanSurface3D and subclasses for different rendering methods which I mostly added because I wanted to test while diagnosing issues with QuadTree and prepare for optional Terrain3D integration. I think there are also plenty of situations when rendering with the simplest PlaneMeshOceanSurface3D makes the most sense, and simplifying the setup of a surface for an end user is useful.

I like this change, but would rename OceanSurface3DBase to just OceanSurface3D to align with the naming in the engine which does not include "Base" in the names of base classes usually.

> Another change is that Ocean3D copies the default material instead of using it directly. There were a lot of issues with 2 Ocean3D's affecting the same resource so I'm not sure it should be possible.

Duplicating the material for each Ocean3D should work.

> I have been thinking maybe it should use the resource directly as a separate "initial settings" property but then use a different material to write its changes to, or the various initial settings on the material should be properties of Ocean3D itself that it handles applying to the material properly? Still thinking about this.

I would say have Ocean3D expose and manage the settings on the material.

> See if creating one node then duplicating using instantiation for each other node per LOD level is cheaper. Not finding much documentation on that, it might only be useful for keeping node changes in-sync

This may be a good candidate for using the RenderingServer directly to draw the surface [as described here](https://docs.godotengine.org/en/4.0/tutorials/performance/using_servers.html#instantiating-a-mesh-into-3d-space).

> Quick test (not committed) shows that this kind of works but is a bit jittery

Also had the same jitter when I tried this. I suspect there are so many nodes that moving them triggers it.

> Found a bug with the QuadTree implementation that I've fixed here, but let me know if I should PR it separately real quick since it also happens in devel

I've been chasing that for a bit, that's great that you corrected that. I don't think a separate PR is needed.

> There's another issue with changing the camera to the second one when the first is very far away, but I'm not sure it's really possible to fix since I think it has to do with how the simulation simplifies at a distance. Basically all the buoyant objects wind up in the wrong place under or over the surface.

The simulation itself does not simulate at a distance, only the visual rendering does. The buoyancy remains the same "resolution" regardless of distance from the camera. I do notice that the domain warp setting causes the buoyancy bodies to not properly follow the wave height for the location they are in. Does what you are seeing here occur with the domain warp strength set to 0? If that fixes it, the issue is how domain warp is handled in `Ocean3D.global_to_pixel()`.

Tattomoosa committed:
OceanSurface3DBase -> OceanSurface3D

Tattomoosa committed:
Fix QuadTree3D not finding editor camera, remove unused code

Tattomoosa committed:
MultiMeshOceanSurface3D use plane size for instance distance

Tattomoosa committed:
OceanSurface3D infinite rendering options

Tattomoosa committed:
QuadTree3D calculate cull_box in within_sphere

- Issues with infinite ocean rendering settings since cull_box stayed at
  it's original position

Tattomoosa committed:
QuadTree3D: handle null camera

Tattomoosa committed:
OceanSurface3D - reposition during idle time

I think it helps with the jitter when infinite is on a little bit?
More testing required.

Tattomoosa commented:
> I like this change, but would rename OceanSurface3DBase to just OceanSurface3D to align with the naming in the engine which does not include "Base" in the names of base classes usually.

Done

> I would say have Ocean3D expose and manage the settings on the material.

Cool. Thinking that will be part of this PR.

> Also had the same jitter when I tried this. I suspect there are so many nodes that moving them triggers it.

This isn't the cause, even a single plane still displays the jitter. It isn't noticeable at all when the subdivision amount is high enough, so I'm thinking it has to do with the vertices being able to get into the displaced position consistently.

Not sure that's actually solvable but went ahead and implemented basic infinity in the base OceanSurface3D class anyway for further experimentation and because at high enough detail levels its very usable. If the jitter of doing it the simple way isn't solvable, then it can still be done by teleporting meshes around instead of moving them slowly. Probably not going to look at that as part of this PR though.

> This may be a good candidate for using the RenderingServer directly

Agreed. Probably not going to look at that as part of this PR either, but it does interest me and I'd be happy to look into it a little later.

> The simulation itself does not simulate at a distance, only the visual rendering does. The buoyancy remains the same "resolution" regardless of distance from the camera. I do notice that the domain warp setting causes the buoyancy bodies to not properly follow the wave height for the location they are in. Does what you are seeing here occur with the domain warp strength set to 0? If that fixes it, the issue is how domain warp is handled in Ocean3D.global_to_pixel().

Thanks for clearing that up, but it does appear to still happen with domain warp at 0. I don't have any ideas for the cause here, but I do have some ideas for setting up testing scenes that might help to sort it out.

### TODO before PR

>  ~~When Ocean3D editor simulation is on, it should pause when a game instance is running~~

I think I've finally sorted out how to do this, looking at implementing it now. EDIT: Found a much easier way, actually. Done

> ~~QuadTree3D makes all of nodes in-editor but only ever shows the lowest LOD, meaning it can be really slow to open scenes~~

I wish I had a better idea how to handle this, but for now I'm just going to stop the quad tree from making all LOD levels in-editor. I think once this PR is ready to merge I'm going to look at the Terrain3D integration and then I might take a deep dive into the quadtree and try to tackle a couple of the things we've talked about. EDIT: Done

> Ocean3D doesn't handle changing the FFT settings correctly while simulating, maybe others, need to test more

I think it's just the FFT settings that are a problem, but all the Ocean3D code that's dealing with the RenderingServer, compute shaders, etc, is pretty opaque to me. I know the issue is the change in buffer size if the newer FFT size is bigger than the old one. Not sure how to solve that cleanly just yet though.

I'm also getting this error I've had some issue tracking down:

```
E 0:00:00:0580   set_texture_rd_rid: Condition "!RD::get_singleton()->texture_is_valid(p_texture_rd_rid)" is true.
  <C++ Source>   scene/resources/texture_rd.cpp:78 @ set_texture_rd_rid()
```

It seems to happen either not at all, 3 times, or 6 times, when running a scene with Ocean3D which is also simulating Ocean3D within the editor. It looks to me like the error occurs when a texture has an RID and a new RID is assigned to it (?)
https://github.com/godotengine/godot/blob/b09f793f564a6c95dc76acc654b390e68441bd01/scene/resources/texture_rd.cpp#L78

It's definitely behavior introduced by Ocean3D as a tool script, and I assume it's one of the rendering thread textures or compute shaders but the error doesn't stack trace back to GDScript

So for both that error and FFT settings, would appreciate some guidance if you have ideas, or I *think* the PR is set up so that you can push to it directly. I'll work on get the changes I know how to tackle done before I get back to poking at those

Tattomoosa committed:
Ocean3D - do not simulate in-editor if playing scene

Tattomoosa committed:
Merge branch 'devel' into ocean-tool-script

Tattomoosa committed:
QuadTree3D: Do not make subquads in-editor

tessarakkt commented:
Hey! Sorry I fell off the earth for a while, life being life and keeping me away from side projects. Hoping to start getting back into things again.

Testing this now in Godot 4.2.2 and I do not get any ocean rendered in the editor. It runs fine when actually running the scene, but there is no surface rendered. Haven't looked much into this yet.

I'm thinking the buoyancy problem you noticed is likely related to the issue described in #17. Probably a combination of the reduced mesh resolution as you mentioned earlier in the thread that gets made worse by certain settings. For now we'll consider this its own issue to look into.

> I think it's just the FFT settings that are a problem, but all the Ocean3D code that's dealing with the RenderingServer, compute shaders, etc, is pretty opaque to me. I know the issue is the change in buffer size if the newer FFT size is bigger than the old one. Not sure how to solve that cleanly just yet though.

I would say the way to deal with this is to free all of the the image buffer RIDs (stored in the `*_tex_cascade` arrays in Ocean3D) and empty the arrays they were in, then I _think_ `Ocean3D._initialize_simulation()` can be run again without problems (hopefully). The shader handling code in Ocean3D is going to get some attention soon once I move to Godot 4.4, this is going to be one of those changes, so this doesn't need to be worried about here.

    E 0:00:00:0580   set_texture_rd_rid: Condition "!RD::get_singleton()->texture_is_valid(p_texture_rd_rid)" is true.
    <C++ Source>   scene/resources/texture_rd.cpp:78 @ set_texture_rd_rid()

I also get this error. For me it consistently happens 3 times both when loading the scene in the editor that contains an Ocean3D, and when running the scene. I suspect when it happened 6 times for you, you likely had 2 Ocean3D instances in the scene.

Based on the error message, it appears to be related to this section in `Ocean3D._initialize_simulation()`:

	for i in cascade_ranges.size():
		## Initialized empty, it will be generated each frame
		_spectrum_tex_cascade[i] = _rd.texture_create(_fmt_rg32f, RDTextureView.new(), [initial_image_rgf.get_data()])
		_spectrum_uniform_cascade[i] = RDUniform.new()
		_spectrum_uniform_cascade[i].uniform_type = RenderingDevice.UNIFORM_TYPE_IMAGE
		_spectrum_uniform_cascade[i].binding = Binding.SPECTRUM
		_spectrum_uniform_cascade[i].add_id(_spectrum_tex_cascade[i])
		
		## Bind the displacement map cascade texture to the visual shader
		_waves_image_cascade[i] = Image.create(fft_resolution, fft_resolution, false, Image.FORMAT_RGF)
		_waves_texture_cascade[i] = Texture2DRD.new()
		_waves_texture_cascade[i].texture_rd_rid = _spectrum_tex_cascade[i]

What is happening here, for each cascade, is initialize a texture buffer which is used both for calculations done by the compute shaders, and is also the buffer where the resulting displacement maps generated by those shaders end up. These compute shaders run in a separate RenderingDevice than the one used by the engine for rendering, so in order for the main rendering device to be able to access these buffers without copying the data, the Texture2DRD provided by Godot takes the image buffer on the compute device, and makes it available to the main render thread.

The error suggests that the RIDs in `_spectrum_tex_cascade[]` do not point to valid textures during the last line. But the errors appear to be thrown before that code is even reached. The errors also do not appear to interfere with successful rendering.

I'm not quite sure what is causing these yet. None of the code changed in this PR stands out to me as something that might cause that.
